File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_map<char,int> storage;
        
        int n = s.size();
        int l = 0;
        int r = 0;
        int result = 0;
        
        while (l < n && r < n){
            if (storage.find(s[r]) == storage.end()){
                storage[s[r]]++;
                r++;
                result = max(result, r-l);
            } else {
                storage.erase(s[l]);
                l++;
            }
        }
        return result;
    }
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        storage = {}
        l = 0
        r = 0
        out = 0
        while r < len(s):
            if s[r] in storage:
                del storage[s[l]]
                l += 1
            else:
                storage[s[r]] = 1
                r += 1
                out = max(out, r-l)
        return out